corehq/apps/app_manager/management/commands/check_for_xss_attempts.py

```diff
@@ -0,0 +1,75 @@
+import re
+from datetime import datetime
+from couchdbkit.exceptions import ResourceNotFound
+from django.conf import settings
+from django.urls import reverse
+from lxml.html.clean import Cleaner
+
+from .form_iterator import FormIteratorCommandBase, get_current_apps, get_forms
+
+
+class Command(FormIteratorCommandBase):
+
+    def __init__(self):
+        self.js_usage_found = 0
+        self.total_forms = 0
+        self.start_time = None
+        self.iteration_key = __name__
+
+    def process_forms(self, all, limit, reset, batchsize):
+        get_apps = self.get_all_apps if all else get_current_apps
+
+        for app in get_apps(reset, batchsize):
+            for form in get_forms(app):
+                js_usage, value = form_contains_xss_attempt(form)
+                if js_usage:
+                    handle_entity_form(form, app, value, self.log_file)
+                    self.js_usage_found += 1
+                self.total_forms += 1
+                if limit > 0 and self.total_forms >= limit:
+                    return
+
+    def report_results(self):
+        elapsed = datetime.now() - self.start_time
+        self.broadcast(f'Found {self.js_usage_found} JS usages in {self.total_forms} forms')
+        self.broadcast(f'Completed search in {elapsed}')
+
+
+def form_contains_xss_attempt(form):
+    regex = '(?<=<value>)(.*?)(?=\s*<\/value>)'
+    regexoutput = '(?<=<output)(.*?)(?=\s*>)'
+    cleaner = Cleaner()
+    clean_html = cleaner.clean_html
+    cleaner_js = Cleaner(javascript=False, safe_attrs_only=False)
+    cleaner_html_js = cleaner_js.clean_html
+
+    try:
+        source = form.source
+    except ResourceNotFound:
+        return False, ''
+
+    for value in re.findall(regex, form.source):
+        if '&lt;' and '&gt;' in value:
+            value = value.replace("&lt;", "<").replace("&gt;", ">")
+        if '<output' in value:
+            for output in re.findall(regexoutput, value):
+                value = value.replace(output, '')
+            value = value.replace("<output>", '').replace("</output>", '')
+        cleaned_js = cleaner_html_js("<div>" + value + "</div>")[5:-6]
+        cleaned = clean_html("<div>" + value + "</div>")[5:-6]
```

Comment:
I agree on separating out the part that prepares the extracted value to be put through the cleaner function but I'm not sure if there's another way around using the magic numbers. It's purely there so the string it writes to the log doesn't have the extra `<div>` tags at the ends so I could get rid of it entirely but for what it's worth I've yet to see the cleaner change the added div tags into `<p>`. I actually went with div tags because that's the only thing it won't touch - appending <p> tags around the value actually returns the cleaned html wrapped in div tags.

Reply:
Just to clarify, a magic numbers are:
> Unique values with unexplained meaning or multiple occurrences which could (preferably) be replaced with named constants

the issue is that `5` and `6` have no innate relationship to `<div>` and `</div>` -- the initial code relies on the reader to realize that those values are actually the lengths of the enclosing tags. As soon as those numbers are stored in a named constant that describes their meaning, they cease to be 'magic'